### PR TITLE
Add support to customize the RetryHydrate function

### PR DESCRIPTION
**Pull request by Subhajit97**
```
type RetryConfig struct {
	ShouldRetryErrorFunc ErrorPredicateWithContext
	// deprecated use ShouldRetryErrorFunc
	ShouldRetryError ErrorPredicate

	// Maximum number of retry operation to be performed. Default set to 10.
	MaxAttempts int64
	// Algorithm for the backoff. Supported values: Fibonacci, Exponential, and Constant. Default set to Fibonacci.
	BackoffAlgorithm string
	// Starting interval. Default set to 100ms.
	RetryInterval int64
	// Set a maximum on the duration (in ms) returned from the next backoff.
	CappedDuration int64
	// Sets a maximum on the total amount of time (in ms) a backoff should execute.
	MaxDuration int64
}
```

### CODE REVIEW

File: plugin/hydrate_error.go [hunk 1]
```diff
@@ -47,6 +54,57 @@ func RetryHydrate(ctx context.Context, d *QueryData, hydrateData *HydrateData, h
 	return hydrateResult, err
 }
 
+func getBackoff(retryConfig *RetryConfig) (retry.Backoff, error) {
```
Comment (kaidaguerre) [inline plugin/hydrate_error.go:57]: could you do something like 
```
func getBackoff(retryConfig *RetryConfig) (retry.Backoff, error) {
	// Default set to Fibonacci
	backoffAlgorithm := "Fibonacci"
	retryInterval := time.Duration(100) // in ms

	// Check from config
	if retryConfig != nil {
		if retryConfig.BackoffAlgorithm != "" {
			backoffAlgorithm = retryConfig.BackoffAlgorithm
		}

		if retryConfig.RetryInterval != 0 {
			retryInterval = time.Duration(retryConfig.RetryInterval)
		}
		if retryConfig.CappedDuration != 0 {
			backoffAlgorithm = "CappedDuration"
		}

		if retryConfig.MaxDuration != 0 {
			backoffAlgorithm = "CappedDuration"
		}
	}

	var backoff retry.Backoff
	var err error
	switch backoffAlgorithm {
	case "Fibonacci":
		backoff, err = retry.NewFibonacci(retryInterval * time.Millisecond)
		if err != nil {
			return nil, err
		}
	case "Exponential":
		backoff, err = retry.NewExponential(retryInterval * time.Millisecond)
		if err != nil {
			return nil, err
		}
	case "Constant":
		backoff, err = retry.NewConstant(retryInterval * time.Millisecond)
		if err != nil {
			return nil, err
		}
	case "CappedDuration":
		backoff = retry.WithCappedDuration(time.Duration(retryConfig.CappedDuration)*time.Millisecond, backoff)
	case "MaxDuration":
		backoff = retry.WithMaxDuration(time.Duration(retryConfig.MaxDuration)*time.Second, backoff)
	}

	return backoff, nil
}
```

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: plugin/hydrate_error.go (hunk 1 above)
Comment (Subhajit97) [reply plugin/hydrate_error.go:57]: @kaidaguerre `CappedDuration` and `MaxDuration` are used to apply caps or limits on the backoff algorithms.
If we use the following code
```
if retryConfig.CappedDuration != 0 {
	backoffAlgorithm = "CappedDuration"
}

if retryConfig.MaxDuration != 0 {
	backoffAlgorithm = "CappedDuration"
}
```
It will directly go to either `case "CappedDuration":` or `case "MaxDuration":`, which is not correct.

File: plugin/hydrate_error.go [hunk 2]
```diff
@@ -47,6 +54,54 @@ func RetryHydrate(ctx context.Context, d *QueryData, hydrateData *HydrateData, h
...
+	// Default set to Fibonacci
+	backoffAlgorithm := "Fibonacci"
+	retryInterval := time.Duration(100) // in ms
```
Comment (kaidaguerre) [inline plugin/hydrate_error.go:60]: you could always rephrase this to 
```
retryInterval := 100 * time.Millisecond
```
to avoid the need for a comment

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Review by kaidaguerre — APPROVED** [review]
